File: Weekly/week_09/Keepbang/HappyNumber.java
class Solution {
    public boolean isHappy(int n) {
        return isHappyFn(n,new ArrayList());
    }

    public boolean isHappyFn(int n, ArrayList arr){
        if(n == 1){
            return true;
        }

        int sum = sumFn(n);
        for(int i = 0;i<arr.size();i++){
            if(arr.get(i).equals(sum)){
                return false;
            }
        }

        arr.add(sum);

        return isHappyFn(sum,arr);
    }
    public int sumFn(int n){
        return Arrays.stream(Stream.of(String.valueOf(n).split(""))
                .mapToInt(Integer::parseInt).toArray())
                .reduce(0,(sum, num) -> sum + (num*num));
    }
}

File: Weekly/week_09/Keepbang/HappyNumber.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
    return isHappyFn(n,[]);
};

var isHappyFn = function(n,arr){
    if(n == 1){
        return true;
    }

    let sum = sumFn(n);
    for(let i = 0;i<arr.length;i++){
        if(arr[i] == sum){
            return false;
        }
    }

    arr.push(sum);

    return isHappyFn(sum,arr);
}

var sumFn = function(n){
    return n.toString().split("").reduce((sum,n)=> sum += Math.pow(n,2),0);
}